### PR TITLE
Improve pilot analysis to only analyze changed configs during analysis intervals

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Fix https://github.com/istio/istio/issues/30200

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [x] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Each analyzer may depend on all other types, since we have cross-type references in analysis? So its not clear to me how this works.

**Comment by hanxiaop** [comment]
> Each analyzer may depend on all other types, since we have cross-type references in analysis? So its not clear to me how this works.

@howardjohn I've added a new function `ForEachNeedsAnalyze` which is similar to `ForEach`, the difference is that `ForEach` iterates over all of the configs of a type, but `ForEachNeedsAnalyze` only iterates over configs of a type that needs to be analyzed in the next analysis run. 
There are two situations here:
1. Resources of a type are interrelated. If so, if there is any config change to a specific resource, all of the configs in that type will be re-analyzed again. 
2. Resources depends on other types. For this one, just use `ForEach` and every configs of that type will be obtained.

File: pkg/config/analysis/analyzers/envoyfilter/envoyfilter.go [hunk 1]
```diff
@@ -45,13 +45,20 @@ func (*EnvoyPatchAnalyzer) Metadata() analysis.Metadata {
...
 // Analyze implements analysis.Analyzer
 func (s *EnvoyPatchAnalyzer) Analyze(c analysis.Context) {
+	var hasConfigChange bool
 	// hold the filter names that have a proxyVersion set
-	patchFilterNames := make([]string, 0)
-	c.ForEach(collections.IstioNetworkingV1Alpha3Envoyfilters.Name(), func(r *resource.Instance) bool {
-		names := s.analyzeEnvoyFilterPatch(r, c, patchFilterNames)
-		patchFilterNames = names
-		return true
+	c.ForEachNeedsAnalyze(collections.IstioNetworkingV1Alpha3Envoyfilters.Name(), func(r *resource.Instance) bool {
```
Comment (howardjohn) [inline pkg/config/analysis/analyzers/envoyfilter/envoyfilter.go:50]: should we make a function for this like `c.TypeChange()` or something?

Comment (howardjohn) [reply pkg/config/analysis/analyzers/envoyfilter/envoyfilter.go:50]: Or make it automatic actually. `analysis.Metadata` already has `Inputs` field

File: pkg/config/analysis/analyzers/gateway/gateway.go [hunk 2]
```diff
@@ -50,7 +50,7 @@ func (*IngressGatewayPortAnalyzer) Metadata() analysis.Metadata {
 
 // Analyze implements analysis.Analyzer
 func (s *IngressGatewayPortAnalyzer) Analyze(c analysis.Context) {
-	c.ForEach(collections.IstioNetworkingV1Alpha3Gateways.Name(), func(r *resource.Instance) bool {
+	c.ForEachNeedsAnalyze(collections.IstioNetworkingV1Alpha3Gateways.Name(), func(r *resource.Instance) bool {
```
Comment (howardjohn) [inline pkg/config/analysis/analyzers/gateway/gateway.go:53]: This (and other places) doesn't really seem safe. The analyzer actually depends on Gateway+Pod+Services interaction, but we only run it on Gateway change. Really if the Service changes, we should update it as well

**Review by howardjohn — COMMENTED** [review]
I really feel the entire architecture of analyzer is not fixable. It was fundamentally designed in a non-scaleable way....

**Comment by hanxiaop** [comment]
> I really feel the entire architecture of analyzer is not fixable. It was fundamentally designed in a non-scaleable way....

@howardjohn Agree. Just look into the code, and it looks like it's hard to deal with specific config changes. I'll talk with the UX group to see if there are any ideas of the refactorization. 

> This (and other places) doesn't really seem safe. The analyzer actually depends on Gateway+Pod+Services interaction, but we only run it on Gateway change. Really if the Service changes, we should update it as well

For this one, I've made some improvements, but it's not competely fix the issue. The idea is to watch config changes, however instead of analyzing the changed configs, we analyze the entire list of resources of that type, and also lists of other types which are related to this type in the analysis. By doing this, at least we won't waste resources on some almost-never-changed configs.

File: pkg/config/analysis/analyzers/gateway/gateway.go (hunk 2 above)
Comment (hanxiaop) [reply pkg/config/analysis/analyzers/gateway/gateway.go]: Revised, now if service A has changed, the entire lists of resources from Gateway, Pod and Service collections will be analyzed to entire the statuses are always synced.

**Pull request closed without merging**